### ISSUE TITLE
Provide a nicer syntax for tuples

### ISSUE THREAD

**Issue by wereHamster** [purescript/purescript#1687]
Especially 3 or more element 'tuples'. Creating those is reasonably easy with `Data.Tuple.Nested`, but pattern matching on them is a nightmare. It adds so much noise into the code. Is there a reason why PureScript can't support the Haskell syntax?

**Comment by garyb** [comment]
This has come up a few times before. @paf31's idea was that records make tuples mostly redundant, and additionally are better for code comprehension since things have to be explicitly named rather than implied by position.

Over time, I have found situations where I miss the syntactic convenience of tuples when doing something like a `zip` or whatever, where tuples are just useful as a temporary data structure and naming isn't so important as their use is an internal implementation detail, but now we have record puns that has been alleviated somewhat too. Also, unlike tuples, you already get accessors for any part of a record for free, as you can use `_.foo` to access any field, you're not restricted to `fst` and `snd` for 2-tuples.

So... yeah. I can take them or leave them at this point, not sure what @paf31 thinks nowadays. Just thought I'd explain why they haven't been present from the start, and would suggest maybe records would work better if you're moving past a 2-tuple in arity anyway.

**Comment by wereHamster** [comment]
compare

```haskell
x = case (mbA, eitherB, head []) of
    (Nothing, Right b, Nothing) -> ...
    (Nothing, _,       Just h ) -> ...
    (Just a,  _,       _      ) -> ...
```

```haskell
x = case { mbA: mbA, eitherB: eitherB, someHead: head [] } of
    { mbA: Nothing, eitherB: Right b, someHead: Nothing } -> ...
    { mbA: Nothing, eitherB: _,       someHead: Just h  } -> ...
    { mbA: Just a,  eitherB: _,       someHead: _       } -> ...
```

**Comment by garyb** [comment]
Yeah, that's a good example of the kind of case I had in mind where tuples are handy as a temporary structure. I haven't had to write code like that frequently enough that it has become a major annoyance personally.

The specific construction you have there can be made slightly simpler with record puns (`{ mbA, eitherB, someHead: head [] }`), but that obviously isn't always the case, and it doesn't help with the destructuring either.

Another thing that might help with this is pattern synonyms (#1667), as they could be provided for the `Nested` tuples then too.

**Comment by wereHamster** [comment]
Another example to consider: Often I want to use such an anonymous structure as key in a Map or another structure which requires certain instances on the type. But that's not possible because PureScript doesn't automatically derive Eq and Ord. So I have to define a local newtype and manually write the instances. This adds a lot of noise to the code.

**Comment by user471** [comment]
It would be nice, if you could use record pattern matching without fields.
```
x = case { mbA: mbA, eitherB: eitherB, someHead: head [] } of
    { Nothing, Right b, Nothing } -> ...
    { Nothing, _,   Just h } -> ...
    { Just a,  _,  _ } -> ...
```

**Comment by garyb** [comment]
You can already do that.

``` purescript
x = case { mbA: mbA, eitherB: eitherB, someHead: head [] } of
    { mbA: Nothing, eitherB: Right b, someHead: Nothing } -> ...
    { mbA: Nothing, someHead: Just h } -> ...
    { mbA: Just a } -> ...
```

As mentioned in an issue you raised previously, records are a set of label/value pairs rather than a list, so omitting the labels is not going to be possible.

**Comment by user471** [comment]
> records are a set of label/value pairs rather than a list
Any record has order in declaration, so why not use it?

**Comment by zudov** [comment]
@user471 No they don't. Record is an *unordered* labeled set, in other words `{ a: 3, b: 4 }` is equivalent to `{ b: 4, a: 3 }`. Of course syntactically we have some ordering (as you said "in declaration") but this ordering information is lost immediately after the declaration happens.

**Comment by user471** [comment]
> Record is an unordered labeled set,

Why cannot record be ordered labeled set?

> { a: 3, b: 4 } is equivalent to { b: 4, a: 3 }

Yes, because name would have more priority than position.

> information is lost immediately after the declaration happens

Than it just should be saved.

**Comment by zudov** [comment]
I think what we rather want are type operators (https://github.com/purescript/purescript/issues/416). The we would be able to have something like:

```purescript
data a :*: b = a :*: b

foo :: (a :*: b :*: c) -> d

case (a :*: b :*: c) of
  a :*: b :*: c -> foo (a :*: b :*: c) 
```

**Comment by zudov** [comment]
@user471 

> Why cannot record be ordered labeled set?

The example that you gave already collides with existing syntax
```purescript
case { a: 1, b: 2 } of
   { a, b } -> a + b
```
But lets assume that we don't yet have such syntax (or we come up with some other syntax to pattern match records based on ordering).

Would that typecheck, if records are ordered?
```purescript 
foo :: { a :: Int, b :: Int }
foo = { b: 1, a: 2 }
```

If it would, then the following expression would also typecheck, but fail at runtime:
```haskell
case { a: Maybe Int, b: Int } of
    { b, Just a } -> a + b
```

What if it wouldn't typecheck? Well, we break a lot of code (not that it's a dealbreaker, but still). But we should remember that our record type `{ a :: A, b :: B}` is simply a sugar for `Object (a :: A, b :: B)`, which means that to make records ordered we would have to make rows ordered too. I didn't think about it in details, but I guess that making rows ordered would make current extensible effect system problematic.

**Comment by garyb** [comment]
Please open a new issue if you want to further discuss the set/list behaviour of records, but here's one last reason not to do it: what is the inferred type of `{ 1, "foo" }`?

**Comment by user471** [comment]
```
case { a: Maybe Int, b: Int } of
    { b, Just a } -> a + b
```
No, it would not typecheck. Why would compiler swap positions in ``` { a: Maybe Int, b: Int } ```?

**Comment by user471** [comment]
>what is the inferred type of { 1, "foo" }?

Why would it need a type? It's just a syntax for pattern matching.

Although, it could have a type https://github.com/purescript/purescript/issues/1267

**Comment by zudov** [comment]
@user471 Let's not continue a discussion irrelevant to an issue that we are in. I'll answer your points, if you open another issue or ping me on gitter/irc.

**Comment by user471** [comment]
https://github.com/purescript/purescript/issues/1690

**Comment by paf31** [comment]
> It would be nice, if you could use record pattern matching without fields.

But then you wouldn't be able to reorder the fields in the input object since that would change the meaning of the binders. More generally, it ties the meaning of the binders to the values they bind, which is undesirable.

**Comment by paf31** [comment]
I still think records trump anonymous tuple types, especially now with record puns. The one downside is a lack of type class instances, but you get those back by newtyping.

**Comment by user471** [comment]
>But then you wouldn't be able to reorder the fields in the input object since that would change the meaning of the binders.

But how often do you reorder the input object fields? 
Anyway, it only matter if you use this feature, if you want to be able to reorder fields, just don't use it.
You cannot just reorder tuple, but everyone use it.

**Comment by jprider63** [comment]
I'm not sure what the status of this is, but I just wanted to say I'd prefer to have this syntax. Here's an example where I think this syntax would improve the code: 

    helper (Tuple (Tuple pre post) false) c = tuple3 pre (post <> [c]) false
    helper (Tuple (Tuple pre post) true) c =
        if f c then
            tuple3 (pre <> [c]) post true
        else
            tuple3 pre (post <> [c]) false

**Comment by hdgarrood** [comment]
I think this is resolved by #1696. The examples @wereHamster gave near the start can be written much simpler now.

There's also the `uncurryN` functions in `Data.Tuple.Nested`, which would make examples like @jprider63's a lot nicer:

```purescript
helper c = uncurry3 go
  where
  go pre post false = tuple3 pre (post <> [c]) false
  go pre post true =
    if c
      then tuple3 (pre <> [c]) post true
      else tuple3 pre (post <> [c]) false
```

So I think we should close this for now.

**Comment by paf31** [comment]
Agreed.
